api/database/datasource_repo.py

```diff
@@ -51,3 +52,31 @@ async def datasource_exists(account, container) -> bool:
         True if the datasource exists, False otherwise.
     """
     return await get(account, container) is not None
+
+
+async def create(datasource: DataSource) -> DataSource:
+    """
+    Create a new datasource. The datasource will be henceforth identified by account/container which
+    must be unique or this function will return a 400.
+    This will encrypt the sasToken if it is provided.
+
+    Parameters
+    ----------
+    datasource : DataSource
+        The datasource to create.
+
+    Returns
+    -------
+    DataSource
+        The datasource.
+    """
+    datasource_collection: AgnosticCollection = collection()
+    if await datasource_exists(datasource.account, datasource.container):
+        raise Exception("Datasource Already Exists")
+    if datasource.sasToken:
```

Comment:
Really like this extraction - is there any potential future need to do this with the update as well?

Reply:
I do think it would make sense to add this to others to avoid keeping the logic in the route itself but I would say one thing at time.